libraries/libfc/src/string.cpp

```diff
@@ -1,98 +1,136 @@
-
-     if (itr != str.end() || !fc::is_valid_utf8( str )) {
-        str = escape_string(str, nullptr, escape_ctrl == escape_control_chars::on);
-        modified = true;
-        if (str.size() > max_len) {
-           str.resize(max_len);
-           truncated = true;
-        }
-     }
-
-     if (truncated) {
-        str += add_truncate_str;
-     }
-
-     return std::make_pair(std::ref(str), modified);
-  }
-
-
-} // namespace fc
+   } catch (const boost::bad_lexical_cast& e) {
+      FC_THROW_EXCEPTION(parse_error_exception, "Couldn't parse double");
+   }
+   FC_RETHROW_EXCEPTIONS(warn, "${i} => double", ("i", i))
+}
+
+std::pair<std::string&, bool> escape_str(std::string& str, escape_control_chars escape_ctrl, std::size_t max_len,
+                                         std::string_view add_truncate_str) {
+   bool modified = false, truncated = false;
+   // truncate early to speed up escape
+   if (str.size() > max_len) {
+      str.resize(max_len);
+      modified = truncated = true;
+   }
+   auto itr = escape_ctrl == escape_control_chars::on
+                 ? std::find_if(str.begin(), str.end(),
+                                [](const auto& c) {
+                                   return c == '\x7f' || c == '\\' || c == '\"' || (c >= '\x00' && c <= '\x1f');
+                                })
+                 : std::find_if(str.begin(), str.end(), [](const auto& c) {
+                      // x09 = \t, x0a = \n,                   x0d = \r
+                      return c == '\x7f' || (c >= '\x00' && c <= '\x08') || c == '\x0b' || c == '\x0c' ||
+                             (c >= '\x0e' && c <= '\x1f');
+                   });
+
+   if (itr != str.end() || !fc::is_valid_utf8(str)) {
+      str      = escape_string(str, nullptr, escape_ctrl == escape_control_chars::on);
+      modified = true;
+      if (str.size() > max_len) {
+         str.resize(max_len);
+         truncated = true;
+      }
+   }
+
+   if (truncated) {
+      str += add_truncate_str;
+   }
+
+   return std::make_pair(std::ref(str), modified);
+}
+
+std::vector<std::string> split(const std::string& str, char delim, std::size_t max_split) {
+   std::vector<std::string> out;
+   std::string delim_str{&delim, 1};
+   for (auto&& token : std::views::split(str, delim)) {
+      std::string_view token_str{token.begin(), token.end()};
+      if (max_split > 0 && out.size() >= max_split) {
+         out.back()
+            .append(delim_str)
+            .append(token_str.data(), token_str.size());
+      } else {
+         out.emplace_back(token_str);
+      }
+   }
+   return out;
+}
+
+std::string to_lower(std::string s) {
+   std::ranges::transform(
+      s, s.begin(),
+      [](unsigned char c) {
+         return std::tolower(c);
+      }
+      );
+   return s;
+}
+
+
+
+bool all_digits(const std::string_view& s) {
+   return all_digits(std::string{s.data(),  s.size()});
+};
```

Comment:
Hard to tell in this diff, but there is an `all_digits` above. Shouldn't create a copy of the `string_view` here.